csharp/src/Google.Protobuf.Test/FieldMaskTreeTest.cs

```diff
@@ -380,6 +380,20 @@ public void Merge(bool useDynamicMessage)
                 clearedSource, destination, options, useDynamicMessage);
             Assert.IsNull(destination.Payload);
 
+            // Test merging unset source message fields with inner field paths.
```

Comment:
I don't think I've seen these tests before... I'll split them up into separate actual tests when we're done with these PRs. (It's fine for this PR not to do that.)

Reply:
I'll just move this test to another PR, as it seems to be an unrelated issue. I *do* think that this is correct, because otherwise you will not be able to induce this behavior with settings (whereas with this behavior, you can set `options.ReplacePrimitiveFields = true` and get the other "always overwrite" behavior)

Additionally, `FieldMask` defines the exhaustive list of subproperties as equivalent to just specifying the parent property, which this would violate (specifying just parent property would result in null not overwriting, whereas specifying all properties explicitly would change that).